debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java

```diff
@@ -347,6 +354,81 @@ protected void handleEvent(Event event) {
         }
     }
 
+    protected void bufferEvent(Event event) {
+        if (event == null) return;
+        if (event.getHeader().getEventType() == EventType.QUERY) {
+            QueryEventData command = unwrapData(event);
+            logger.debug("Received query command: {}", event);
+            String sql = command.getSql().trim();
+            if (sql.equalsIgnoreCase("BEGIN")) {
+                beginTransaction(event);
+            } else if (sql.equalsIgnoreCase("COMMIT")) {
+                completeTransaction(true, event);
+            } else if (sql.equalsIgnoreCase("ROLLBACK")) {
+                rollbackTransaction();
+            } else {
+                consumeEvent(event);
+            }
+        } else if (event.getHeader().getEventType() == EventType.XID) {
+            completeTransaction(true, event);
+        } else {
+            consumeEvent(event);
+        }
+    }
+
+    private void consumeEvent(Event event) {
+        if (txStarted) {
+            buffer.add(event);
+        } else {
+            handleEvent(event);
+        }
+    }
+
+    private void beginTransaction(Event event) {
+        if (txStarted) {
```

Comment:
Is this a legit case? Or just to be cautious? Does MySQL support nested TX btw. (in which case that implementation wouldn't be correct)?

Reply:
It is just a defensive measure. It is not probable that the situation will happen but in case the commit or XID event would be garbled or lost the buffer will be able to recover.